src/components/FeaturePanel/PublicTransport/requestRoutes.ts

```diff
@@ -29,11 +29,26 @@ const getTagValue = (
   }
   const altElement = routes.find(({ tags }) => tags[key]);
   if (altElement) {
-    return altElement[key];
+    return altElement.tags[key];
   }
   return undefined;
 };
 
+const getService = (tags: Record<string, string>, routes: WithTags[]) => {
+  const getVal = (key: string) => getTagValue(key, tags, routes);
+  const serviceTagValue = getTagValue('service', tags, routes);
+  const serviceTag =
+    serviceTagValue === 'highspeed' ? 'high_speed' : serviceTagValue;
+  const isHighspeed = getVal('highspeed') === 'yes';
+
+  return (
+    serviceTag ||
+    (isHighspeed && 'high_speed') ||
+    getVal('route') ||
+    getVal('route_master')
+  );
+};
```

Comment:
Can you please explain what this should do? I find the code very hard to read. I guess there could be two things mixed together - like getting a tag from relations `tags` and getting it from the `routes[].tags`. When is which scenario used?

Maybe then we can find more suitable code.

Reply:
It always looks for the relations `tags` and then `routes[].tags`, so like this:
1. `tags.service`
2. `routes[].tags.service`
3. `tags.highspeed=yes`
4. `routes[].tags=highspeed=yes`
5. `tags.route`
6. `routes[].tags.route`
7. ...

Reply:
Ok, I understand now. So, the semantics of `getTagValue()` is in reality `get-tag-or-get-tag-on-child-route`. I think this function does two things which are unrelated.

More readable code for me would be:

```
const getValue = (
  key: string,
  tags: Record<string, string>,
) => {
  if (tags[key]) {
    return tags[key];
  }
  return undefined;
};

const getValueOnAnyRoute = (
  key: string,
  routes: WithTags[],
) => {
  const altElement = routes.find(({ tags }) => tags[key]);
  if (altElement) {
    return altElement.tags[key];
  }
  return undefined;
};

const getService = (tags: Record<string, string>, routes: WithTags[]) => {
  const getVal = (key: string) => getValue(key, tags) ?? getValueOnAnyRoute(key, routes);
```

Do you get the difference? It is subtle, but code like this makes it easier for next reader to get the meaning. They can only read the `const getVal` line to get the full meaning, without opening the complicated (boilerplate) accessing code.

Of course, this is more like pseudo-code highlight the idea, it could be probably refactored much more.

If you want, you can edit it in some followup.